src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c

```diff
@@ -0,0 +1,727 @@
+      }
+      else
+      {
+          timeInfo->inputBufferAdcTime = ((float) l_lStreamLatency / (float) 1000000);
+      }
+  }
+
+}
+
+
+void PaPulseAudio_StreamRecordCb( pa_stream * s,
+                                  size_t length,
+                                  void *userdata )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) userdata;
+    PaStreamCallbackTimeInfo timeInfo = { 0, 0, 0 };    /* TODO: IMPLEMENT ME */
+    int l_iResult = paContinue;
+    long numFrames = 0;
+    int i = 0;
+    size_t l_lDataSize = 0;
+    size_t l_lLocation = 0;
+    size_t l_lBufferSize = 0;
+    const void *l_ptrSampleData = NULL;
+
+
+    if( l_ptrStream == NULL )
+    {
+        PA_DEBUG( ("Portaudio %s: Stream is null!\n",
+                  __FUNCTION__) );
+        return;
+    }
+
+
+    memset( l_ptrStream->inBuffer,
+            0x00,
+            PULSEAUDIO_BUFFER_SIZE );
+
+    /* Stream ain't active yet.
+     * Don't call callback before we have settled down
+     */
+    if( !l_ptrStream->isActive )
+    {
+        l_iResult = paComplete;
+    }
+    else
+    {
+        PaUtil_BeginCpuLoadMeasurement( &l_ptrStream->cpuLoadMeasurer );
+    }
+
+    PaPulseAudio_updateTimeInfo( s,
+                                 &timeInfo,
+                                 1 );
+
+    /* If stream activated or not.. we have to
+     * read what there is coming to memory
+     */
+    while( pa_stream_readable_size( s ) > 0 )
+    {
+        l_ptrSampleData = NULL;
+
+        if( pa_stream_peek( s,
+                            &l_ptrSampleData,
+                            &l_lDataSize ))
+        {
+            PA_DEBUG( ("Portaudio %s: Can't read audio!\n",
+                      __FUNCTION__) );
+            return;
+
+        }
+
+        if( l_lDataSize > 0 && l_ptrSampleData )
+        {
+            memcpy( l_ptrStream->inBuffer + l_lBufferSize,
+                    l_ptrSampleData,
+                    l_lDataSize );
+            l_lBufferSize += l_lDataSize;
+        }
+        else
+        {
+            PA_DEBUG( ("Portaudio %s: Can't read!\n",
+                      __FUNCTION__) );
+        }
+
+        pa_stream_drop( s );
+    }
+
+    if(l_ptrStream->isActive )
+    {
+        if( l_ptrStream->bufferProcessor.streamCallback != NULL )
+        {
+            PaUtil_BeginBufferProcessing( &l_ptrStream->bufferProcessor,
+                                          &timeInfo,
+                                          0 );
+            PaUtil_SetInterleavedInputChannels( &l_ptrStream->bufferProcessor,
+                                                0,
+                                                l_ptrStream->inBuffer,
+                                                l_ptrStream->inSampleSpec.channels );
+            PaUtil_SetInputFrameCount( &l_ptrStream->bufferProcessor,
+                                       l_lBufferSize / l_ptrStream->inputFrameSize );
```

Comment:
I think we need to make sure only compose a buffer of the requested size by the user application. 
Not sure if this is done magically somewhere else, but this will leads to a latency penalty and should be bypassed anyway.